### PR TITLE
Make Hudi bootstrap table use custom split

### PR DESCRIPTION
**Pull request by codope**
Make Hudi bootstrap table use custom split. 

Implemented custom split converters for Hudi
BootstrapBaseFileSplit and RealtimeBootstrapBaseFileSplit.

Test plan -
Added unit tests to cover util methods and Hive page source provider.
Synced a Hudi bootstrap table to Hive and queried through
Presto to verify the patch.
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by codope** [comment]
@vinothchandar @umehrot2 Can you please review this?

**Comment by rubenssoto** [comment]
@vinothchandar @umehrot2

A Gentle reminder to review this PR :)

Thank you so much

**Comment by arunthirupathi** [comment]
I will get back to you by Friday this week. In the mean time, can you please edit the commit message to 72 char limit.

**Comment by vinothchandar** [comment]
@arunthirupathi thanks for the review and the quick turnarounds! I think the changes should be localized to Hudi tables, but will wait for your review.

**Comment by rubenssoto** [comment]
@arunthirupathi only a gentle reminder.

**Comment by arunthirupathi** [comment]
The current design is hacky. It requires at least a day or two for me to understand this. Currently my office work is keeping me busy and I plan to do this on weekend, but due to personal reasons, I can't spend the time required to wrap my head around it.

**Comment by vinothchandar** [comment]
Thanks for taking the time @arunthirupathi ! Want to set the context straight here. Hudi tables were integrated as special hive tables in Presto, per what was recommended to us at that time by presto maintainers. If we want to revisit this, that's totally fine. but would love for a seamless migration path for users out there and also want to understand why the Hive connector cannot embrace these new table formats, rather than trying to special case them.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java
Comment (arunthirupathi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:201]: This property is enabled by default, so by default this would change the behavior rather than opt in ? Was that the intention ?

Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:201]: Yes. We thought that keeping it enabled by default strikes a balance between seamless deployment and mitigating risk. If the reader path breaks, one can disable the config.

Comment (vinothchandar) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: If we don't have this by default, queries would return incorrect results by default. It'd be a large issues for many companies out there.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java
Comment (vinothchandar) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:515]: does making this non-static cut down any parameters? Just throwing it out there, given it seems pretty long

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java
Comment (vinothchandar) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java:168]: @arunthirupathi this code was already how it was working. So behavior change wise, this is not new. FYI

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePageSourceProvider.java
Comment (vinothchandar) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePageSourceProvider.java:197]: these tests should help catch any subtle change in behavior

**Review by vinothchandar — COMMENTED** [review]
Changes seem safe to me.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java
Comment (arunthirupathi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:414]: This method can return Optional.empty(), Is that an acceptable return value ?

**Review by arunthirupathi — APPROVED** [review]
Left some nits, once addressed, I will merge this in.

**Comment by arunthirupathi** [comment]
As discussed offline, There is no other code using the shouldUseRecordReaderFromInputFormat, so we are going to enable the record cursor by default. But if we found out later, this breaks some other connector, we have to revert and re do it.

File: presto-hive/src/main/java/com/facebook/presto/hive/util/HudiBootstrapBaseFileSplitConverter.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/util/HudiBootstrapBaseFileSplitConverter.java:61]: The if-condition guards this. So, we enter this block only when `CUSTOM_FILE_SPLIT_CLASS_KEY` is there and it matches a specific class. That class key is populated together with the other keys in `extractCustomSplitInfo` method.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java:1119]: Good catch. Thanks. I'll rename it in the next commit.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:414]: Yes, that was the behavior even before. If the page source is empty then we throw exception in the caller:
```
if (pageSource.isPresent()) {
    return pageSource.get();
}
throw new IllegalStateException("Could not find a file reader for split " + hiveSplit);
```

**Comment by vinothchandar** [comment]
Thanks @arunthirupathi !

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java
Comment (pratyakshsharma) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:484]: @codope so I was just trying to understand the reason behind calling this method again here. This is already called on line 414 after checking some preconditions. Is it really needed to call it here? Your inputs are appreciated. :)

Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:484]: `getPageSourceFromCursorProvider` just extracts out the common processing logic and is called again here for completeness (no missing return statement). The call on line 414 is there so as to ignore page source factories if a custom record reader is to be used.

**Pull request merged**
